megalista_dataflow/uploaders/google_ads/customer_match/abstract_uploader.py

```diff
@@ -14,206 +14,212 @@
-    if len(destination.destination_metadata) >= 5 and len(destination.destination_metadata[4]) > 0:
-      return destination.destination_metadata[4].replace('-', '')
-    return account_config.google_ads_account_id.replace('-', '')
-
-  @utils.safe_process(logger=logging.getLogger(_DEFAULT_LOGGER))
-  def process(self, batch: Batch, **kwargs) -> None:
-    if not self.active:
-      logging.getLogger(_DEFAULT_LOGGER).warning(
-          'Skipping upload to ads, parameters not configured.')
-      return
-
-    execution = batch.execution
-
-    self._assert_execution_is_valid(execution)
-
-    customer_id = self._get_customer_id(execution.account_config, execution.destination)
-
-    # get API services
-    offline_user_data_job_service = self._get_offline_user_data_job_service(
-      customer_id)
-
-    list_resource_name = self._create_list_if_it_does_not_exist(
-      customer_id, execution.destination.destination_metadata[0],
-      self.get_list_definition(
-        execution.account_config,
-        execution.destination.destination_metadata))
-
-    rows = self.get_filtered_rows(
-      batch.elements, self.get_row_keys())
-
-    # Upload is divided into 3 parts:
-    # 1. Create Job
-    # 2. Create operations (data insertion)
-    # 3. Run the Job
-
-    # 1. Create Job
-    job_creation_payload = {
-      'type_': 'CUSTOMER_MATCH_USER_LIST',
-      'customer_match_user_list_metadata': {
-        'user_list': list_resource_name
-      }
-    }
-
-    job_resource_name = offline_user_data_job_service.create_offline_user_data_job(customer_id = customer_id, job = job_creation_payload).resource_name
-
-    # 2. Create operations (data insertion)
-    operator = self._get_list_operator(execution.destination.destination_metadata[1])
-    data_insertion_payload = {
-      'resource_name': job_resource_name,
-      'enable_partial_failure': False,
-      'operations': [{
-        operator: {
-          'user_identifiers': rows
+    def _assert_execution_is_valid(self, execution) -> None:
+        destination = execution.destination.destination_metadata
+
+        # The number of parameters vary by upload. This test could be parameterized
+        if not destination[0]:
+            raise ValueError('Missing destination information. Received {}'.format(
+                str(destination)))
+
+    def _get_customer_id(self, account_config: AccountConfig, destination: Destination) -> str:
+        """
+          If the customer_id is present on the destination, returns it, otherwise defaults to the account_config info.
+        """
+        if len(destination.destination_metadata) >= 5 and len(destination.destination_metadata[4]) > 0:
+            return destination.destination_metadata[4].replace('-', '')
+        return account_config.google_ads_account_id.replace('-', '')
+
+    def _get_job_by_list_name(self, offline_user_data_job_service, list_resource_name: str, customer_id: str) -> str:
+        if list_resource_name in self._job_cache:
```

Comment:
@astivi It looks like this cache doesn't separate remove and addition jobs.
The docs mentions: "Do not mix create and remove operations within the same OfflineUserDataJob. Doing so can result in a CONFLICTING_OPERATION error." 
in https://developers.google.com/google-ads/api/docs/remarketing/audience-types/customer-match#:~:text=do%20not%20mix%20create%20and%20remove%20operations%20within%20the%20same%20offlineuserdatajob.%20doing%20so%20can%20result%20in%20a%20conflicting_operation%20error.